### DISCUSSION TITLE
x-bind expression.trim is not a function

### DISCUSSION THREAD

**Discussion by marcreichel** [alpinejs/alpine#3734] [7. Bugs (archived)] [answered]
When updating to v3.13.0 `x-bind` expressions are broken somehow.

The error:

<img width="635" alt="image" src="https://github.com/alpinejs/alpine/assets/7645035/c2b68bf6-3fe5-46d2-aae2-6f9f78bf3ff3">

Everything works as expected though but the console is getting flooded with errors.
This happens on all `x-bind` statements.

Downgrading to v3.12.3 fixes the issue, so it's definitely a problem in v3.13.0.

**Comment by ekwoka** [comment]
Can you show a specific x-bind that causes this?

Because I can't think of any valid x-binds that wouldn't have an expression.

So an example that fails would be very helpful for seeing why this might be an issue

**Reply by marcreichel** [reply]
I have some `x-bind` statements (as described [here](https://alpinejs.dev/globals/alpine-bind)) that get added programmatically via:

```javascript
element.setAttribute("x-bind", "DropdownMenu");
```

which results in the error described above. As said in v3.12.3 this is working just fine.

**Reply by ekwoka** [reply]
What does `DropdownMenu` then bind? because just doing `setAttribute` and an expression that resolves to an object works...

**Reply by marcreichel** [reply]
I've set up an `DropdownMenu` with:

```javascript
Alpine.bind("DropdownMenu", () => ({ /* ... */ }))
```

And then bind an element via `element.setAttribute("x-bind", "DropdownMenu")`. The element itself is generated on the fly as well like this:

```javascript
const element = document.createElement("div");
element.setAttribute("x-bind", "DropdownMenu");

document.body.appendChild(element);
```

Doing this results in the error described above.

**Reply by SimoTod** [reply]
It must be something else in your code, see https://codepen.io/SimoTod/pen/WNLwwxN
Likely something already had a bug in your code but the error was not reported.
The trim was added on the last release, somehow something that should have been a string made it to that part of the code.
Hard to say without a reproducible example.

**Reply by ekwoka** [reply]
> Doing this results in the error described above.

But you left out the part I was asking about...

```js
({ /* ... */ })
```

what's in that? That's where the bug is.

**Reply by marcreichel** [reply]
@ekwoka

```javascript
Alpine.bind("DropdownMenu", ({
    "x-data"() {
        return {
            show: false,
            element: this.$el,
            items: this.$el.__dropdownItems, // gets provided from outside
            activeIndex: null,
            init() {
                this.$nextTick(() => {
                    this.show = true;
                });
            },
        };
    },
    "x-show"() {
        return this.show;
    },
    "x-transition:enter": "transition ease-out duration-100",
    "x-transition:enter-start": "opacity-0 scale-90",
    "x-transition:enter-end": "opacity-100 scale-100",
    ":class"() {
        return ["some", "classes"];
    },
    ":style"() {
        return { some: "styles" };
    },
    "x-trap.inert": true,
}));

const element = document.createElement("div");
element.__dropdownItems = [/* array of dropdown items */];
element.setAttribute("x-bind", "DropdownMenu");

document.body.appendChild(element);
```

Sadly, I cannot see, where in the code it would lead to a non-string expression tbh.

The `DropdownMenu` is an example here. I have multiple “components” that use `x-bind` dynamically like described above and all of them have the same issue.

**Reply by marcreichel** [reply]
But it sure seems to be something on my side that's not working correctly. As the pen of @SimoTod reflects exactly how my components are getting rendered/added to the page.

I will look into this and downgrade to 3.12.3 for now.

**Reply by ekwoka** [reply, marked as answer]
@marcreichel 

The x-trap.inert

Values have to be string or function. Can't be boolean.

`true.trim()` is not a function

**Reply by marcreichel** [reply]
@ekwoka omg... that was the problem indeed... I thought using a boolean just adds the attributes. My bad.

Thank you very much!

**Reply by ekwoka** [reply]
It was, before, basically being treated as the string `"true"` so it was working, but was technically out of spec.

Maybe it would be worth doing the conversion explicitly for books and nums
